File: 0359-logger-rate-limiter/0359-logger-rate-limiter.cpp
#include <unordered_map>
using namespace std;
class Logger {
    unordered_map<string, int> times;
public:
    Logger() {
    }
    
    bool shouldPrintMessage(int timestamp, string message) {
        if ((times.find(message) != times.end()) && (timestamp < times[message] + 10)) {
            return false;
        }

        times[message] = timestamp;
        return true;
    }
};

/**
 * Your Logger object will be instantiated and called as such:
 * Logger* obj = new Logger();
 * bool param_1 = obj->shouldPrintMessage(timestamp,message);
 */

File: 0359-logger-rate-limiter/0359-logger-rate-limiter.py
class Logger:

    def __init__(self):
        self.q = collections.deque() # (timestamp, msg)
        self.msg = set()
        

    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        
        while self.q:
            t, m = self.q[0]
            if t > (timestamp - 10):
                break
            self.q.popleft()
            self.msg.remove(m)

        if message in self.msg:
            return False
        else:
            self.q.append((timestamp, message))
            self.msg.add(message)
            return True


        


# Your Logger object will be instantiated and called as such:
# obj = Logger()
# param_1 = obj.shouldPrintMessage(timestamp,message)